keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1241 @@
+| Pod                                       | status.nominatedNodeName           | string           | status.nominatedNodeName       |                                                               |
+| Secret                                    | type                               | enum             | type                           | string(enum)                                                  |
+
+[^1]: This field selector is commented with: "This is a bug, but we need to support it for backward compatibility.".
+
+Note that:
+
+- Resources always provide a value for a selectable field. Notice that in the
+  above table, *all* field selectors for optional fields provide a fallback if the
+  field is omitted.
+- Field selectors only exist for strings, enums, integers and booleans.
+- Field selectors for name and namespace are provided automatically.
+- Field selectors that do not match the path name of the actual field do exist,
+  for historical reasons, but the best practice is for field selectors to match
+  the exact path name of the field.
+
+Also note that we do not currently publish any field selector information into
+OpenAPIv3, and we do not document in APIs or on the Kubernetes website which
+fields may be selected. Because of this, finding the field selectors available for built-in types, to construct the above table, required
+locating [`GetAttrs/*ToSelectableFields` functions in the Kubernetes source code](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#L340).
+
+### Plan
+
+#### API
+
+```go
+type CustomResourceDefinitionVersion struct {
+  
+  // selectableFields specifies paths to fields that may be used as field selectors.
+  // TODO: A max limit will be decided during API review
+	// See https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors
+	SelectableFields []SelectableField `json:"selectableFields,omitempty" protobuf:"bytes,9,rep,name=selectableFields"`
+}
+
+  // SelectableField specifies the JSON path of a field that may be used with field selectors.
+type SelectableField struct {
+	// jsonPath is a simple JSON path which is evaluated against each custom resource to produce a
+  // field selector value.
+  // Only JSON paths without the array notation are allowed.
+  // Must point to a field of type string, boolean or integer. Types with enum values
+  // and strings with formats are allowed.
+  // If jsonPath refers to absent field in a resource, the jsonPath evaluates to an empty string.
+  // Must not point to metdata fields.
+	JSONPath string `json:"jsonPath" protobuf:"bytes,1,opt,name=jsonPath"`
+}
+```
+
+Note that we do not offer a way to set a default value. We discussed this in API
+review and observed that because OpenAPI schema already provides a way to set
+a default value for a field, that we don't need (or want) to add a second way to set a
+default for selectable fields.  Instead, if the jsonPath refers to an absent field
```

Comment:
I assume this means that if a default value is configured and the client doesn't set the field we will end up with the default value for the selectable field

(probably pretty clear, just wanted to confirm)

Reply:
Correct. I've rephrased this to try to make it a bit more clear.